File: python/design/snake.py
import math

import drawsvg as draw

steps = 100
head_size = 80
max_dist = 400
revolutions = 2


def make_circle(max_steps):
    def inner(i):
        frac = i / max_steps

        radius = head_size * frac
        color = f"hsl({frac * 360}, 90%, 50%)"

        angle = 2 * math.pi * revolutions * frac
        x = math.cos(angle) * max_dist * frac
        y = math.sin(angle) * max_dist * frac

        return draw.Circle(x, y, radius, fill=color)

    return inner


def make_snake():
    g = draw.Group()
    for c in map(make_circle(steps), range(1, steps + 1)):
        g.append(c)
    return g


d = draw.Drawing(1000, 1000, origin="center")
d.append(make_snake())
d.save_svg("snake.svg")
print("Saved snake.svg")

File: haskell/design/snake.hs
import Data.Colour.RGBSpace (uncurryRGB)
import Data.Colour.RGBSpace.HSL (hsl)
import Data.Colour.SRGB (sRGB)
import Diagrams.Backend.SVG (B, renderSVG)
import Diagrams.Prelude

steps, revolutions :: Int
steps = 100
revolutions = 2

headSize, maxDist :: Double
headSize = 80
maxDist = 400

-- make_circle(max_steps)(i): Circle(radius) >> fill(color) >> move(x, y)
makeCircle :: Int -> Int -> Diagram B
makeCircle maxSteps i =
  circle radius # fc color # lw none # translate (r2 (x, y))
 where
  frac = fromIntegral i / fromIntegral maxSteps
  radius = headSize * frac
  color = uncurryRGB sRGB (hsl (frac * 360) 0.9 0.5)
  angle = 2 * pi * fromIntegral revolutions * frac
  x = cos angle * maxDist * frac
  y = sin angle * maxDist * frac

-- make_snake(): Group(map(make_circle(steps))([1..steps]))
makeSnake :: Diagram B
makeSnake = mconcat (map (makeCircle steps) [1 .. steps])

main :: IO ()
main = do
  renderSVG "snake.svg" (mkWidth 1000) makeSnake
  putStrLn "Saved snake.svg"
